accepted/cranelift-isel-isle-peepmatic.md

```diff
@@ -0,0 +1,926 @@
+    ;; extractor) on `x`:
+    (rule (EntryPoint x) (A x))
+```
+  
+- When during pattern-matching we see an internal term that does not
+  statically resolve, we need to run any defined rules "in reverse".
+  An example might serve to illustrate:
+  
+```plain
+    (rule (A x 1) (B x))
+    (rule (A x 2) (C x))
+    
+    ;; The `(B x)` pattern, because `B` is an internal term and not
+    ;; an extractor, will search for any rule that produces `B` at the
+    ;; root, then substitute the pattern for that rule (here `(A x 1)`)
+    ;; at that point in the pattern.
+    (rule (EntryPoint (B x)) x)
+```
+
+  In essence, one can see an internal term as being equivalent to (i)
+  an extractor and (ii) a constructor definition, each of whose bodies
+  is composed of logic collected from (i) all rules that produce this
+  internal term, and (ii) all rules that match on this internal term,
+  respectively. In fact, this is a valid (though possibly inefficient)
+  compilation strategy.
+  
+The above implies that defining a rule with `(rule ...)` may attach an
+"internal extractor" to an internal term, if one is produced at the
+root on the right-hand side, and/or an "internal constructor" to an
+internal term, if one is matched at the root on the left-hand side. If
+the left-hand side pattern's *root* symbol is an externally-defined
+extractor, then this means that it will only ever be evaluated in then
+reverse direction. Similarly, if the right-hand side expression's
+*root* symbol is an externally-defined constructor, then this means it
+will only ever be evaluated in the forward direction. An example of
+the former -- a rule that has an external extractor at the pattern
+root root -- follows:
+    
+```plain
+
+    (decl extern (Extractor ...) ...)
+    (extractor Extractor "extractor")
+    
+    (decl (A ...) ...)
+    
+    (rule (Extractor a) (A a))
+    
+    (decl (EntryPoint ...) ...)
+    
+    ;; This will inline `A` in the pattern-matching phase, running
+    ;; "in reverse", bottoming out in a call to `extractor`.
+    (rule (EntryPoint (A a)) a)
+```
+  
+# Compilation Strategy
+
+In order to compile ISLE to Rust code, we perform several steps.
+
+## Single-Pass Elaboration
+
+A traditional term-rewriting system typically works in an iterative
+fashion: rewrite rules are applied, editing the representation
+(e.g. expression tree) in memory, until no more rules apply. This
+*dynamic scheduling* of rewrite rules is costly, as is the actual
+reification of the intermediate states. In contrast, we expect this
+DSL to be compiled into a single-pass transform, and we add certain
+restrictions to rules to enable this.
+
+The basic strategy is to "inline" the rule-matching that would apply
+to the newly constructed value on the right-hand side of a rule. In
+other words, if a rule transforms `(A x y)` to `(B x y)` and there is
+another rule that matches on `(B x y)`, we generate code that chains
```

Comment:
Yes, the DSL itself is agnostic to whatever the input and output formats are (that depends on the "bindings" one writes with external extractor and constructor definitions), so in principle one could have all of the above.

Reductions in the legalization sense, e.g. lowering a 64-bit add to two 32-bit adds (to take a simple example), should be possible as well. This has implications for the term stratification approach to ensuring termination, so we would need to expand that a bit (the RFC notes that allowing substructural recursion, or maybe just "recurse during inlining up to limit k") to allow the lowering code to be generated properly. But that's absolutely a goal -- writing fallback rules like this then composing with machine-specific rules is much more powerful than the support-all-features-on-all-machines approach we currently have (e.g. adding i128 separately on x64 and aarch64).

We could also have multiple passes; @fitzgen and I discussed this a bit yesterday actually. For now I think it makes sense to start with the single lowering pass, because we have the existing machinery for that, but once the DSL compiler exists, it should be relatively straightforward to use it to generate another pass as well.